drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,412 @@
+ *
+ * The Read RX Buffer instruction provides a means to quickly address a receive
+ * buffer for reading. This instruction reduces the SPI overhead by one byte,
+ * the address byte. The command byte actually has four possible values that
+ * determine the address pointer location.
+ *
+ * The Load TX Buffer instruction eliminates the eight-bit address required by
+ * a normal write command. The eight-bit instruction sets the address pointer
+ * to one of six addresses to quickly write to a transmit buffer that points to
+ * the “ID” or “data” address of any of the three transmit buffers.
+ *
+ * @param[in]  id           filter identifier in the MCP2515 mailbox
+ * @param|out] bytebuf      buffer instruction
+ */
+static void fill_standard_id(uint32_t id, uint8_t *bytebuf);
+
+/**
+ * @brief Fill tx/rx extended buffer instruction from filter identifier @p id
+ *
+ * for more details see fill_standard_id.
+ *
+ * @param[in]  id           filter identifier in the MCP2515 mailbox
+ * @param|out] bytebuf      buffer instruction
+ */
+static void fill_extended_id(uint32_t id, uint8_t *bytebuf);
+
+int mcp2515_init(candev_mcp2515_t *dev, void (*irq_handler_cb)(void *))
+{
+    int res;
+
+    res = gpio_init_int(dev->conf->int_pin, GPIO_IN_PU, GPIO_FALLING,
+                        (gpio_cb_t)irq_handler_cb, (void *)dev);
+    if (res != 0) {
+        DEBUG("Error setting interrupt pin!\n");
+        return -1;
+    }
+    gpio_init(dev->conf->rst_pin, GPIO_OUT);
+
+    res = mcp2515_spi_init(dev);
+    if (res < 0) {
+        return -1;
+    }
+
+    uint8_t cmd = MCP2515_RXB0CTRL_MODE_RECV_ALL;
+
+    res = mcp2515_spi_write(dev, MCP2515_RXB0CTRL, &cmd, 1);
+    if (res < 0) {
+        DEBUG("failed to set acceptance mode\n");
+        return -1;
+    }
+    return 0;
+}
+
+void mcp2515_reset(candev_mcp2515_t *dev)
+{
+    gpio_clear(dev->conf->rst_pin);
+    xtimer_usleep(RESET_DELAY_US);
+    gpio_set(dev->conf->rst_pin);
+    xtimer_usleep(osc_startup(dev));
+}
+
+static void fill_standard_id(uint32_t id, uint8_t *bytebuf)
+{
+    bytebuf[0] = (uint8_t)((id & 0x000007F8UL) >> 3);   /* T/RXBnSIDH */
+    bytebuf[1] = (uint8_t)((id & 0x00000007UL) << 5);   /* T/RXBnSIDL */
+    bytebuf[2] = (uint8_t)((id & 0xFF000000UL) >> 24);  /* T/RXBnEID8 */
+    bytebuf[3] = (uint8_t)((id & 0x00FF0000UL) >> 16);  /* T/RXBnEID0 */
+}
+
+static void fill_extended_id(uint32_t id, uint8_t *bytebuf)
+{
+    bytebuf[0] = (uint8_t)((id & 0x1FE00000UL) >> 21);          /* T/RXBnSIDH */
+    bytebuf[1] = (uint8_t)((id & 0x001C0000UL) >> 13)
+                 | (uint8_t)((id & 0x00030000UL) >> 16) | 0x08; /* T/RXBnSIDL */
+    bytebuf[2] = (uint8_t)((id & 0x0000FF00UL) >> 8);           /* T/RXBnEID8 */
+    bytebuf[3] = (uint8_t)(id & 0x000000FFUL);                  /* T/RXBnEID0 */
+}
+
+int mcp2515_send(candev_mcp2515_t *dev, const struct can_frame *frame,
+                 int mailbox)
+{
+    uint8_t prio = 1; /* TODO: adjust priority */
```

Comment:
Should this be a define?

Reply:
I don't think it can be a define, because it is passed as a pointer.